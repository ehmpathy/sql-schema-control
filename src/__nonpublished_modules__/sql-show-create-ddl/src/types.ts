export interface DatabaseConnection {
  query: (args: { sql: string; values?: (string | number)[] }) => Promise<{ rows: any[] }>;
  end: () => Promise<void>;
}
