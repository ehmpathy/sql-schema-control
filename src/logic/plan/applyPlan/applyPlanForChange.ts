import { mysql as prepare } from 'yesql';
import { DefinitionPlan, DatabaseConnection, ChangeDefinition } from '../../../types';

/*
  1. run the sql
  3. save the change into change log
*/
export const applyPlanForChange = async ({
  connection,
  plan,
}: {
  connection: DatabaseConnection;
  plan: DefinitionPlan;
}) => {
  // 1. run the sql
  try {
    await connection.query({ sql: plan.definition.sql });
  } catch (error) {
    throw new Error(`Could not apply ${plan.definition.path}: ${error.message}`);
  }

  // 2. upsert that we have applied this change; TODO: adapter pattern to support resources in addition to changes
  await connection.query(
    prepare(`
    INSERT INTO schema_control_change_log
      (change_id, change_hash, change_content)
    VALUES
      (:change_id, :change_hash, :change_content)
    ON DUPLICATE KEY UPDATE
      change_hash = :change_hash,
      change_content = :change_content;
  `)({
      change_id: (plan.definition as ChangeDefinition).id,
      change_hash: (plan.definition as ChangeDefinition).hash,
      change_content: plan.definition.sql,
    }),
  );
};
