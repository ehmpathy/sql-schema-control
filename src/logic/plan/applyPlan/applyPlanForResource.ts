import {
  RequiredAction,
  DefinitionPlan,
  DatabaseConnection,
  ResourceDefinition,
  ResourceType,
} from '../../../domain';

const REAPPLIABLE_RESOURCES = [
  ResourceType.FUNCTION,
  ResourceType.PROCEDURE,
  ResourceType.VIEW,
]; // these resources don't persist data
export const applyPlanForResource = async ({
  connection,
  plan,
}: {
  connection: DatabaseConnection;
  plan: DefinitionPlan;
}) => {
  // grab the resource
  if (plan.definition.constructor !== ResourceDefinition)
    throw new Error('must be resource');
  const resource = plan.definition as ResourceDefinition;

  // 1. if we need to reapply, try to drop the resource first
  if (plan.action === RequiredAction.REAPPLY) {
    if (!REAPPLIABLE_RESOURCES.includes(resource.type))
      throw new Error(`resource ${plan.id} is not reappliable`); // if we can't drop this resource (i.e., its a table) throw an error
    await connection.query({ sql: `DROP ${resource.type} ${resource.name}` });
  }

  // 2. apply it
  try {
    await connection.query({ sql: resource.sql });
  } catch (error) {
    throw new Error(
      `Could not apply ${plan.definition.path}: ${error.message}`,
    );
  }
};
