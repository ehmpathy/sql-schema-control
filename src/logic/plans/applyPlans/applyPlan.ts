import { mysql as prepare } from 'yesql';
import { RequiredAction, DefinitionPlan, DatabaseConnection } from '../../../types';

/*
  1. throw error if RequiredAction is not APPLY or REAPPLY
  2. run the sql
  3. save the change into change log
*/
export const applyPlan = async ({ connection, plan }: { connection: DatabaseConnection, plan: DefinitionPlan }) => {
  // 1. throw an error if required action is not apply or reapply
  if (![RequiredAction.APPLY, RequiredAction.REAPPLY].includes(plan.action)) throw new Error(`plan.action must be ${RequiredAction.APPLY} or ${RequiredAction.REAPPLY}. Got ${plan.action}`);

  // 2. run the sql
  try {
    await connection.query({ sql: plan.definition.sql });
  } catch (error) {
    throw new Error(`Could not apply ${plan.definition.path}: ${error.message}`);
  }

  // 3. upsert that we have applied this change; TODO: adapter pattern to support resources in addition to changes
  await connection.query(prepare(`
    INSERT INTO schema_control_change_log
      (change_id, change_hash, change_content)
    VALUES
      (:change_id, :change_hash, :change_content)
    ON DUPLICATE KEY UPDATE
      change_hash = :change_hash,
      change_content = :change_content;
  `)({
    change_id: plan.definition.id,
    change_hash: plan.definition.hash,
    change_content: plan.definition.sql,
  }));
};
