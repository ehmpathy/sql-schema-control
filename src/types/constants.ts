export enum DefinitionType {
  RESOURCE = 'resource',
  CHANGE = 'change',
}

export enum DatabaseLanguage {
  MYSQL = 'mysql',
}

export interface DatabaseConnection {
  query: (args: { sql: string, values?: any[] }) => Promise<any>;
  end: () => Promise<void>;
}

export enum ChangeDefinitionStatus {
  NOT_APPLIED = 'NOT_APPLIED',
  UP_TO_DATE = 'UP_TO_DATE',
  OUT_OF_DATE = 'OUT_OF_DATE',
}
