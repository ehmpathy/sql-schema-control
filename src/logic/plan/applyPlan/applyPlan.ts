import {
  DatabaseConnection,
  DefinitionPlan,
  ChangeDefinition,
  ResourceDefinition,
} from '../../../types';
import { applyPlanForChange } from './applyPlanForChange';
import { applyPlanForResource } from './applyPlanForResource';

export const applyPlan = async ({
  connection,
  plan,
}: {
  connection: DatabaseConnection;
  plan: DefinitionPlan;
}) => {
  if (plan.definition.constructor === ChangeDefinition)
    return applyPlanForChange({ connection, plan });
  if (plan.definition.constructor === ResourceDefinition)
    return applyPlanForResource({ connection, plan });
  throw new Error('unexpected definition type');
};
