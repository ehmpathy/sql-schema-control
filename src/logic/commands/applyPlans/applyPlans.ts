import { RequiredAction, DefinitionPlan, DatabaseConnection } from '../../../types';
import { applyPlan } from '../../plan/applyPlan';
import { getColoredPlanTitle } from '../utils/getColoredPlanTitle';
import chalk from 'chalk';

/*
  if APPLY or REAPPLY, apply
  else, skip
*/
export const applyPlans = async ({
  connection,
  plans,
}: {
  connection: DatabaseConnection;
  plans: DefinitionPlan[];
}) => {
  // add padding to output
  console.log(''); // tslint:disable-line

  // apply each plan
  for (const plan of plans) {
    // if no change, dont even mention it
    if (plan.action === RequiredAction.NO_CHANGE) continue;

    // if cant apply, notify we're skipping
    const canApply = [RequiredAction.APPLY, RequiredAction.REAPPLY].includes(plan.action);
    if (!canApply) {
      console.log(`  ${chalk.bold(chalk.yellow('↓'))} ${getColoredPlanTitle({ plan })}`); // tslint:disable-line no-console
      continue;
    }

    // if can apply, then try to apply
    try {
      await applyPlan({ connection, plan });
      console.log(`  ${chalk.bold(chalk.green('✔'))} ${getColoredPlanTitle({ plan })}`); // tslint:disable-line no-console
    } catch (error) {
      console.log(`  ${chalk.bold(chalk.red('x'))} ${getColoredPlanTitle({ plan })}\n`); // tslint:disable-line no-console
      console.log(error.message); // tslint:disable-line no-console
      throw error;
    }
  }
};
