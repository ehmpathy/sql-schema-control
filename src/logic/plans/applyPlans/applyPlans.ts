import Listr from 'listr';
import { RequiredAction, DefinitionPlan, DatabaseConnection } from '../../../types';
import { applyPlan } from './applyPlan';
import { getColoredActionToken } from '../_utils/getColoredActionToken';

/*
  if APPLY or REAPPLY, apply
  else, skip
*/
export const applyPlans = async ({ connection, plans }: { connection: DatabaseConnection, plans: DefinitionPlan[] }) => {
  const tasks = plans
    .filter(plan => [RequiredAction.APPLY, RequiredAction.REAPPLY].includes(plan.action)) // only apply or reapply
    .map((plan): Listr.ListrTask => ({
      title: `${getColoredActionToken({ action: plan.action })} ${plan.definition.path}`,
      task: () => applyPlan({ connection, plan }),
    }));
  const taskSuite = new Listr(tasks);
  await taskSuite.run().catch((err) => {
    console.error(err.message);
  });
};
