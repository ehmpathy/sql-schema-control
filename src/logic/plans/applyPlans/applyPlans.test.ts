import sha256 from 'simple-sha256';
import { applyPlans } from './applyPlans';
import { applyPlan } from './applyPlan';
import { ChangeDefinition, DefinitionPlan, DefinitionType, RequiredAction } from '../../../types';
import { stdout } from 'stdout-stderr';

jest.mock('./applyPlan');
const applyPlanMock = applyPlan as jest.Mock;
const sleep = (ms: number) => new Promise(resolve => setTimeout(resolve, ms));
applyPlanMock.mockImplementation(() => sleep(500));

const exampleDefinition = new ChangeDefinition({
  id: '__UUID__',
  type: DefinitionType.CHANGE,
  path: '__PATH__',
  sql: '__SQL__',
  hash: sha256.sync('__SQL__'),
});
const plan = new DefinitionPlan({
  definition: exampleDefinition,
  difference: '__APPLY_DIFFERENCE__',
  action: RequiredAction.APPLY,
});
const noChangePlan = new DefinitionPlan({
  definition: exampleDefinition,
  difference: '__NO_CHANGE_DIFFERENCE__',
  action: RequiredAction.NO_CHANGE,
});
const reapplyPlan = new DefinitionPlan({
  definition: exampleDefinition,
  difference: '__REAPPLY_DIFFERENCE__',
  action: RequiredAction.REAPPLY,
});
const manualReapplyPlan = new DefinitionPlan({
  definition: exampleDefinition,
  difference: '__MANUAL_REAPPLY_DIFFERENCE__',
  action: RequiredAction.MANUAL_REAPPLY,
});

process.stdout.isTTY = true;
describe('applyPlans', () => {
  beforeEach(() => jest.clearAllMocks());
  it('should attempt to apply each plan', async () => {
    await applyPlans({ connection: {} as any, plans: [plan, plan, reapplyPlan] });
    expect(applyPlanMock.mock.calls.length).toEqual(3);
    expect(applyPlanMock.mock.calls[0][0]).toMatchObject({ // check correct input
      plan,
    });
  });
  it('should not attempt to apply NO_CHANGE or MANUAL_REAPPLY plans', async () => {
    await applyPlans({ connection: {} as any, plans: [plan, noChangePlan, manualReapplyPlan, reapplyPlan] });
    expect(applyPlanMock.mock.calls.length).toEqual(2);
  });
  it('should display an expected listr output for the plans', async () => {
    process.stdout.isTTY = undefined; // since listr acts differently if nonTTY and jest is nonTTY when more than one test is run
    stdout.stripColor = false;
    stdout.start();
    await applyPlans({ connection: {} as any, plans: [plan, reapplyPlan] });
    stdout.stop();
    const output = stdout.output.split('\n').filter(line => !line.includes('console.log')).join('\n') // strip the console log portion
      .replace(/\[\d\d:\d\d:\d\d\]/g, ''); // remove all timestamps, since they change over time...
    expect(output).toMatchSnapshot();
    process.stdout.isTTY = true;
  });
});
