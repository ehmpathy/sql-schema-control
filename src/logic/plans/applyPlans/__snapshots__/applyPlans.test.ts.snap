// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`applyPlans should display an expected listr output for the plans 1`] = `
"  [33mâ ™[39m [1m[32m[APPLY][39m[22m __PATH__
    [1m[33m[REAPPLY][39m[22m __PATH__
[2K[1A[2K[1A[2K[G  [33mâ ¹[39m [1m[32m[APPLY][39m[22m __PATH__
    [1m[33m[REAPPLY][39m[22m __PATH__
[2K[1A[2K[1A[2K[G  [33mâ ¸[39m [1m[32m[APPLY][39m[22m __PATH__
    [1m[33m[REAPPLY][39m[22m __PATH__
[2K[1A[2K[1A[2K[G  [33mâ ¼[39m [1m[32m[APPLY][39m[22m __PATH__
    [1m[33m[REAPPLY][39m[22m __PATH__
[2K[1A[2K[1A[2K[G  [32mâœ”[39m [1m[32m[APPLY][39m[22m __PATH__
  [33mâ ™[39m [1m[33m[REAPPLY][39m[22m __PATH__
[2K[1A[2K[1A[2K[G  [32mâœ”[39m [1m[32m[APPLY][39m[22m __PATH__
  [33mâ ¹[39m [1m[33m[REAPPLY][39m[22m __PATH__
[2K[1A[2K[1A[2K[G  [32mâœ”[39m [1m[32m[APPLY][39m[22m __PATH__
  [33mâ ¸[39m [1m[33m[REAPPLY][39m[22m __PATH__
[2K[1A[2K[1A[2K[G  [32mâœ”[39m [1m[32m[APPLY][39m[22m __PATH__
  [33mâ ¼[39m [1m[33m[REAPPLY][39m[22m __PATH__
[2K[1A[2K[1A[2K[G  [32mâœ”[39m [1m[32m[APPLY][39m[22m __PATH__
  [33mâ ´[39m [1m[33m[REAPPLY][39m[22m __PATH__
[2K[1A[2K[1A[2K[G  [32mâœ”[39m [1m[32m[APPLY][39m[22m __PATH__
  [32mâœ”[39m [1m[33m[REAPPLY][39m[22m __PATH__
"
`;
